monitoring/modules.go

```diff
@@ -0,0 +1,209 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var errors []error
+	for _, module := range r.Modules {
+		if !modules.HasLoaded(module) {
+			errors = append(errors, trace.NotFound("module %q not loaded", module))
+		}
+	}
+
+	if len(errors) == 1 {
+		return trace.Wrap(errors[0])
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// ModuleChecker checks if the specified set of kernel modules ara loaded
+type ModuleChecker struct {
+	// Modules lists required kernel modules
+	Modules []string
+	moduleGetter
+}
+
+// LoadModules loads list of kernel modules from /proc/modules
+func LoadModules() (Modules, error) {
+	f, err := os.Open("/proc/modules")
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	defer f.Close()
+
+	modules, err := LoadModulesFrom(f)
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	return modules, nil
+}
+
+// LoadModulesFrom loads list of kernel modules from the specified reader
+func LoadModulesFrom(r io.Reader) (modules Modules, err error) {
+	s := bufio.NewScanner(r)
+
+	for s.Scan() {
+		line := s.Text()
+		module, err := parseModule(line)
+		if err != nil {
+			log.Warnf("failed to parse module, skip line %q", line)
+			continue
+		}
+		modules = append(modules, *module)
+	}
+
+	if s.Err() != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	return modules, nil
+}
+
+// HasLoaded determines whether module names are loaded.
+// names should not have duplicates
+func (r Modules) HasLoaded(names ...string) bool {
+L:
+	for _, module := range r {
+		for i, name := range names {
+			if module.Name == name && module.IsLoaded() {
+				names = append(names[:i], names[i+1:]...)
+				continue L
+			}
+		}
+	}
+	return len(names) == 0
+}
+
+// Modules lists kernel modules
+type Modules []Module
+
+// IsLoaded determines if this module is loaded
+func (r Module) IsLoaded() bool {
+	return r.ModuleState == ModuleStateLive
+}
+
+// Module describes a kernel module
+type Module struct {
+	// ModuleState specifies the state of the module: live, loading/unloading
+	ModuleState
+	// Name identifies the module
+	Name string
+	// Instances specifies the number of instances this module has loaded
+	Instances int
+	// DependsOn lists module's dependencies on other modules
+	DependsOn []string
+}
+
+// parseModule parses module information from a single line of /proc/modules
+// https://www.centos.org/docs/5/html/Deployment_Guide-en-US/s1-proc-topfiles.html#s2-proc-modules
+func parseModule(moduleS string) (*Module, error) {
+	columns := strings.SplitN(moduleS, " ", len(moduleColumns))
+	if len(columns) != len(moduleColumns) {
+		return nil, trace.BadParameter("invalid input: expected six whitespace-separated columns, but got %q",
+			moduleS)
+	}
+
+	instanceS := columns[2]
+	instances, err := strconv.ParseInt(instanceS, 10, 32)
+	if err != nil {
+		return nil, trace.BadParameter("invalid instances field: expected integer, but got %q", instanceS)
+	}
+
+	var dependencies []string
```

Comment:
you are not really using dependencies anywhere, why implement the code?

Reply:
If this is used as a library (and not restricted to use by just the kernel module checker), it may be useful to callers - it is just a speculative implementation.

Reply:
so the answer is we don't need it, so why keep this code :)